File: javascript/src/method/toSet.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"

import {CollectionConstants}                             from "../CollectionConstants"
import {__uniqueValues, __uniqueValuesByArray, __values} from "./_tables utility"
import {isArray}                                         from "./isArray"
import {isArrayByStructure}                              from "./isArrayByStructure"
import {isCollectionHolder}                              from "./isCollectionHolder"
import {isCollectionHolderByStructure}                   from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}                    from "./isMinimalistCollectionHolder"

//#region -------------------- Facade method --------------------

/**
 * Convert the {@link collection} to an {@link ReadonlySet set}
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array}) to convert
 * @extensionFunction
 */
export function toSet<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>,): ReadonlySet<T> {
    if (collection == null)
        return CollectionConstants.EMPTY_SET
    if (isCollectionHolder(collection,))
        return toSetByCollectionHolder(collection,)
    if (isArray(collection,))
        return toSetByArray(collection,)
    if (isMinimalistCollectionHolder(collection,))
        return toSetByMinimalistCollectionHolder(collection,)

    if (isCollectionHolderByStructure<T>(collection,))
        return toSetByCollectionHolder(collection,)
    if (isArrayByStructure<T>(collection,))
        return toSetByArray(collection,)
    return toSetByMinimalistCollectionHolder(collection,)
}


/**
 * Convert the {@link collection} to an {@link ReadonlySet set}
 *
 * @param collection The {@link Nullable nullable} {@link MinimalistCollectionHolder collection} to convert
 * @extensionFunction
 */
export function toSetByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>,): ReadonlySet<T> {
    if (collection == null)
        return CollectionConstants.EMPTY_SET

    const size = collection.size
    if (size == 0)
        return CollectionConstants.EMPTY_SET
    return __withDuplicate(collection, size,)
}

/**
 * Convert the {@link collection} to an {@link ReadonlySet set}
 *
 * @param collection The {@link Nullable nullable} {@link CollectionHolder collection} to convert
 * @extensionFunction
 */
export function toSetByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>,): ReadonlySet<T> {
    if (collection == null)
        return CollectionConstants.EMPTY_SET
    if (collection.isEmpty)
        return CollectionConstants.EMPTY_SET
    if (collection.hasDuplicate)
        return __withDuplicate(collection, collection.size,)
    return __withoutDuplicate(collection, collection.size,)
}

/**
 * Convert the {@link collection} to an {@link ReadonlySet set}
 *
 * @param collection The {@link Nullable nullable} {@link ReadonlyArray collection} to convert
 * @extensionFunction
 */
export function toSetByArray<const T, >(collection: Nullable<readonly T[]>,): ReadonlySet<T> {
    if (collection == null)
        return CollectionConstants.EMPTY_SET

    const size = collection.length
    if (size == 0)
        return CollectionConstants.EMPTY_SET
    return __withDuplicateByArray(collection, size,)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop method --------------------

function __withDuplicate<const T, >(collection: MinimalistCollectionHolder<T>, size: number,) {
    return Object.freeze(new Set(__uniqueValues(collection, size,),),)
}

function __withDuplicateByArray<const T, >(collection: readonly T[], size: number,) {
    return Object.freeze(new Set(__uniqueValuesByArray(collection, size,),),)
}


function __withoutDuplicate<const T, >(collection: MinimalistCollectionHolder<T>, size: number,) {
    return Object.freeze(new Set(__values(collection, size,),),)
}

//#endregion -------------------- Loop method --------------------

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/ToSet.java
package joookiwi.collection.java.method;

import java.util.Set;
import java.util.function.Function;
import java.util.function.Supplier;
import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.callback.ObjIntFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import joookiwi.collection.java.extended.ArrayAsSet;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CollectionConstants.emptySet;
import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;

@NotNullByDefault
public final class ToSet
        extends UtilityWithTable {

    @Contract(ALWAYS_FAIL_0)
    private ToSet() { throw new ImpossibleConstructionException("The utility class “ToSet” cannot be constructed.", ToSet.class); }

    //#region -------------------- Facade methods --------------------

    //#region -------------------- ∅ --------------------

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> @Unmodifiable Set<T> toSet(final @Nullable MinimalistCollectionHolder<? extends T> collection) {
        if (collection == null)
            return emptySet();

        final var size = collection.size();
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> @Unmodifiable Set<T> toSet(final @Nullable CollectionHolder<? extends T> collection) {
        if (collection == null)
            return emptySet();
        if (collection.isEmpty())
            return emptySet();
        if (collection.hasDuplicate())
            return new ArrayAsSet<>(_values(collection, collection.size()));
        return new ArrayAsSet<>(_uniqueValues(collection, collection.size()));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param <T>        The `collection` type
    @ExtensionFunction
    public static <T extends @Nullable Object> @Unmodifiable Set<T> toSet(final T @Nullable @Unmodifiable [] collection) {
        if (collection == null)
            return emptySet();

        final var size = collection.length;
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size));
    }

    //#endregion -------------------- ∅ --------------------
    //#region -------------------- (T, int) → U --------------------

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                                      final ObjIntFunction<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.size();
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size, transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable CollectionHolder<? extends T> collection,
                                                                                                      final ObjIntFunction<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();
        if (collection.isEmpty())
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, collection.size(), transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final T @Nullable @Unmodifiable [] collection,
                                                                                                      final ObjIntFunction<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.length;
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size, transform));
    }

    //#endregion -------------------- (T, int) → U --------------------
    //#region -------------------- (T) → U --------------------

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                                      final Function<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.size();
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size, transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable CollectionHolder<? extends T> collection,
                                                                                                      final Function<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();
        if (collection.isEmpty())
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, collection.size(), transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final T @Nullable @Unmodifiable [] collection,
                                                                                                      final Function<? super T, ? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.length;
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection, size, transform));
    }

    //#endregion -------------------- (T) → U --------------------
    //#region -------------------- () → U --------------------

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                                      final Supplier<? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.size();
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(size, transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final @Nullable CollectionHolder<? extends T> collection,
                                                                                                      final Supplier<? extends U> transform) {
        if (collection == null)
            return emptySet();
        if (collection.isEmpty())
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(collection.size(), transform));
    }

    /// Convert the `collection` to an [immutable][Unmodifiable] [Set]
    /// applying a transformation
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param transform  The given transform
    /// @param <T>        The `collection` type
    /// @param <U>        The new type
    @ExtensionFunction
    public static <T extends @Nullable Object, U extends @Nullable Object> @Unmodifiable Set<U> toSet(final T @Nullable @Unmodifiable [] collection,
                                                                                                      final Supplier<? extends U> transform) {
        if (collection == null)
            return emptySet();

        final var size = collection.length;
        if (size == 0)
            return emptySet();
        return new ArrayAsSet<>(_uniqueValues(size, transform));
    }

    //#endregion -------------------- () → U --------------------

    //#endregion -------------------- Facade methods --------------------

}
